_posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc

```diff
@@ -0,0 +1,2334 @@
+infrastructure code, using a platform that looks something like this:
+
+.Architecture of platform for running Terraform/Terragrunt CI/CD workflows.
+image::/assets/img/guides/infrastructure-cicd-pipeline/tftg-pipeline-architecture.png[]
+
+
+[[use_generic_cicd_platforms_as_a_workflow_engine_but_run_infrastructure_deployments_from_within_your_account]]
+=== Use generic CI/CD platforms as a workflow engine but run infrastructure deployments from within your account
+
+Given the limitations and tradeoffs of the various platforms we covered in <<cicd_platforms>>, we don't recommend
+relying on a single platform for implementing the entire workflow. Instead, we recommend a hybrid solution that takes
+advantage of the strengths of each platform, and cover the weaknesses. The design looks as follows:
+
+- Use any generic CI/CD server you wish to integrate with your VCS repositories.
+- Deploy a self-hosted deployment server within your AWS account that cannot be accessed by anyone but a few admins plus
+  a limited trigger we'll define below.
+- Define your CI workflows so that the CI/CD server triggers deployments against the deploy server.
+
+This design implements separation of the concerns so that we take full advantage of the strengths of each platform,
+while covering the weaknesses: relying on the CI/CD platforms to manage the workflow/pipeline, but have it trigger
+infrastructure deployments on self-hosted systems that are more locked down.
+
+We don't want to give the CI/CD servers permissions to deploy and manage arbitrary infrastructure. CI/CD servers are
+typically not secure enough to handle sensitive information, and you don't want a server that is used for executing
+arbitrary code and regularly used (and written to) by your entire dev team to have admin permissions.
+
+Instead, we delegate this responsibility to an isolated, closed off system in the AWS account that only expose a limited
+set of actions that can be triggered. That way, if anyone gets access to your CI server, they can at most kick off
+builds on existing code, but they don't get arbitrary admin access.
+
+
+[[options_for_deploy_server]]
+=== Options for deploy server
+
+The deploy server needs to be a self-hosted platform in order to satisfy the requirement for isolation. It should also
+avoid executing arbitrary workflows. Finally, it should support configurations options that limit what code can run on
+the server. This limits the options for what you can use as your deploy server. Here is a list of platforms that satisfy
+these constraints, and their strengths and weaknesses:
+
+Gruntwork ECS Deploy Runner Stack::
+  This is a stack you can deploy in your AWS account that sets up an ECS task with a customizable docker container for
+  running `terraform validate`, `terraform plan`, and `terraform apply`, or the Terragrunt equivalent. It is also
+  extensible to support other commands as well, such as running `go test` for Terratest or `packer build` for building
```

Comment:
Perhaps make it clear that it can be extended to run *any* arbitrary command!

Reply:
I think I will keep as is, since that could be seen as a security risk if someone glosses over it.